src/mds/MDSPerfMetricTypes.h

```diff
@@ -0,0 +1,116 @@
+// -*- mode:C++; tab-width:8; c-basic-offset:2; indent-tabs-mode:t -*-
+// vim: ts=8 sw=2 smarttab
+
+#ifndef CEPH_MDS_PERF_METRIC_TYPES_H
+#define CEPH_MDS_PERF_METRIC_TYPES_H
+
+#include <iostream>
+
+#include "include/encoding.h"
+#include "mdstypes.h"
+
+enum UpdateType {
+  UPDATE_TYPE_NOOP = 0,
+  UPDATE_TYPE_REFRESH,
+  UPDATE_TYPE_REMOVE,
+};
+
+// TODO: maybe split out metrics into separate structs so as to
+// forward only those metrics that have changed.
+struct Metrics {
+  // cache utilization percentages
+  double cap_hit_percent = 0.0;
+  double osdc_read_ahead_util_percent = 0.0;
+  // client bandwidth
+  double read_iops = 0.0;
+  double write_iops = 0.0;
+
+  void encode(bufferlist &bl) const {
+    using ceph::encode;
+    ENCODE_START(1, 1, bl);
+    encode(cap_hit_percent, bl);
+    encode(osdc_read_ahead_util_percent, bl);
+    encode(read_iops, bl);
+    encode(write_iops, bl);
+    ENCODE_FINISH(bl);
+  }
+
+  void decode(bufferlist::const_iterator &iter) {
+    using ceph::decode;
+    DECODE_START(1, iter);
+    decode(cap_hit_percent, iter);
+    decode(osdc_read_ahead_util_percent, iter);
+    decode(read_iops, iter);
+    decode(write_iops, iter);
+    DECODE_FINISH(iter);
+  }
+
+  void dump(Formatter *f) const {
+    f->dump_float("cap_hit_percent", cap_hit_percent);
+    f->dump_float("osdc_read_ahead_util_percent", osdc_read_ahead_util_percent);
+    f->dump_float("read_iops", read_iops);
+    f->dump_float("write_iops", write_iops);
+  }
+
+  friend std::ostream& operator<<(std::ostream& os, const Metrics& metrics) {
+    os << "[cap_hit_percent=" << metrics.cap_hit_percent << ", osdc_read_ahead_util_percent="
+       << metrics.osdc_read_ahead_util_percent << ", read_iops=" << metrics.read_iops
+       << ", write_iops=" << metrics.write_iops << "]";
+    return os;
+  }
+};
+WRITE_CLASS_ENCODER(Metrics)
+
+
+
+struct RankMetrics {
+  RankMetrics() { }
+  RankMetrics(mds_rank_t rank) : rank(rank) {
+  }
+
+  mds_rank_t rank = MDS_RANK_NONE;
```

Comment:
Let's also add an sequence number (epoch) here so that rank 0 can determine if another rank is behind. If there is no update, a non-zero rank can still send an update with the new epoch. Likewise, rank 0 needs to send out a heartbeat with a new epoch of its own. `RankMetrics` should include the new sequence number and the last epoch it saw from rank 0.